fix: 脆弱性をなくす

Ryzhon opened the pull request.

Commit:
fix: 脆弱性をなくす

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -4,12 +4,19 @@ class BlogsController < ApplicationController
+
+    return unless @blog.secret && current_user != @blog.user
+
+    raise ActiveRecord::RecordNotFound
```
`ActiveRecord::RecordNotFound` は自分で raise するのではなく Rails の API に raise させてください。

Ryzhon replied:
@blog = Blog.where(secret: false).find_by!(id: params[:id])を追加することでsecretがfalseのものを取得することでエラーが起きるようにしました。

cafedomancer commented on app/helpers/blogs_helper.rb:
```diff
@@ -2,6 +2,7 @@
   def format_content(blog)
-    blog.content.gsub("\n", '<br>').html_safe # rubocop:disable Rails/OutputSafety
+    blog = h(blog.content)
+    blog.gsub("\n", '<br>').html_safe # rubocop:disable Rails/OutputSafety
```
`rubocop:disable` がなくても動くような実装にできないか考えてみてください。

Ryzhon replied:
`simple_format(h(blog.content))`にすることでrubocopを動くようにしました。

cafedomancer commented on app/helpers/blogs_helper.rb:
```diff
@@ -2,6 +2,7 @@
 module BlogsHelper
   def format_content(blog)
-    blog.content.gsub("\n", '<br>').html_safe # rubocop:disable Rails/OutputSafety
+    blog = h(blog.content)
```
`h(blog.content)` はブログそのもの (`blog`) ではないと思います。

cafedomancer commented on app/views/api/liking_users/_user.json.jbuilder:
```diff
@@ -1 +1 @@
-json.extract! user, :id, :email, :created_at, :updated_at, :nickname, :premium
+json.extract! user, :id, :created_at, :updated_at, :nickname, :premium
```
不要な情報のみを削るよりも、必要な情報のみを吐くほうが考え方としてはより安全です。

cafedomancer replied:
対応が漏れています。

Ryzhon replied:
.extract!で必要な情報のみを吐くために`:email`を削除している認識なのですが、extractを変更するということですか？

cafedomancer replied:
email 以外にも不要な情報があるので、判断して必要な情報のみを残してください、ということです。

Ryzhon replied:
`:nickname`のみ返すように修正しました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -28,6 +35,10 @@ def create
   end
 
   def update
+    if !current_user.premium && blog_params[:random_eyecatch] == 'true'
+      redirect_to blog_url(@blog), alert: 'Premium users only can set random eyecatch.'
+      return
+    end
```
要件としては明確に書かれていませんが、無料ユーザーがランダムなアイキャッチを使おうとしていても、ブログ自体の作成・更新はできたほうがいいんじゃないかなーと思います。また、この処理が update にしか書かれていないので、create では脆弱性が引き続き解消されていないような気がしました。

Ryzhon replied:
無料ユーザーがランダムなアイキャッチを使おうとしていても、ブログ自体の作成・更新はできるようにしました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -4,12 +4,19 @@ class BlogsController < ApplicationController
   skip_before_action :authenticate_user!, only: %i[index show]
 
   before_action :set_blog, only: %i[show edit update destroy]
+  before_action :set_edit_blog, only: %i[edit update destroy]
```
`set_blog` と `set_edit_blog` で `Blog` の検索が 2 回行われるケースがあります。データベースへの問い合わせは時間がかかるので、最低限の回数 (ここでは 1 回) で済ませたいです。

Ryzhon replied:
`set_currnet_user_blog`に変更しそれぞれの処理で一回ずつ処理するように修正しました。

Ryzhon committed:
ActiveRecord::RecordNotFoundをAPI側で起こす

Ryzhon committed:
安全でかつ、rubocopが動くように修正

Ryzhon committed:
無料ユーザーがランダムなアイキャッチを使おうとしていても、ブログ自体の作成・更新はできるようにする

Ryzhon committed:
set_currnet_user_blogに変更しそれぞれの処理で一回ずつ処理するようにする

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,18 @@
+  def show
+    return unless @blog.secret && current_user != @blog.user
+
+    @blog = Blog.where(secret: false).find_by!(id: params[:id])
```
https://github.com/Ryzhon/haunted-blog/pull/1#discussion_r1528431499 のコメントと関連しますが、`set_blog` とこの行とで `Blog` の検索が 2 回行われています。

Ryzhon replied:
ユーザーが詳細画面を見るときとブログの所有者でシークレットに設定している詳細画面を見るときでケースを分けなければいけないので、2回実行されてしまうのではないでしょうか。

cafedomancer replied:
1 回の検索で済ませる方法があるので考えてください、ということです。

Ryzhon replied:
一回で検索できるようにしました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,18 @@
   skip_before_action :authenticate_user!, only: %i[index show]
 
-  before_action :set_blog, only: %i[show edit update destroy]
+  before_action :set_blog, only: %i[show]
```
特定のアクションでしか使わないものはフィルタで実装しなくてもよいです。

Ryzhon replied:
削除しました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -28,7 +34,9 @@ def create
   def update
-    if @blog.update(blog_params)
+    updated_params = blog_params
+    updated_params = updated_params.merge(random_eyecatch: false) if !current_user.premium && blog_params[:random_eyecatch] == 'true'
```
明確な要件があるわけではありませんが、以下のようなケースで意図しない更新が行われる可能性があります。意図しない更新は避けたいです。

1. 有料ユーザーで登録
1. ランダムアイキャッチを指定したブログを作成
1. 無料ユーザーに変更
3. 1 で作成したブログを更新 -> ランダムアイキャッチが意図せず無効になる

Ryzhon replied:
おっしゃるような内容まで考慮できていませんでした。
ありがとうございます。
なかなか思いつかなかったため、ひとまずif文で対応しましたがもっといい方法があると考えています。
よろしければご教示いただけると幸いです。

cafedomancer replied:
対応した結果はどこにありますか？

Ryzhon replied:
プッシュし忘れていました。
プッシュしました。

Ryzhon committed:
特定のケースでアイキャッチを更新しないようにする
詳細：
有料ユーザーで登録
ランダムアイキャッチを指定したブログを作成
無料ユーザーに変更
1 で作成したブログを更新 -> ランダムアイキャッチが意図せず無効になる

Ryzhon committed:
必要な項目のみ返すように修正

Ryzhon committed:
fix: showではクエリを一回のみ実行するように修正

cafedomancer commented on app/views/api/liking_users/_user.json.jbuilder:
```diff
@@ -1 +1 @@
-json.extract! user, :id, :email, :created_at, :updated_at, :nickname, :premium
+json.extract! user, :nickname
```
必要な情報も消えてしまっています。必要な情報を残し、その情報が必要だと判断した根拠を示してください。

Ryzhon replied:
テストが通ることを判断の根拠としました。

cafedomancer replied:
修正されていないです。現状でもテストが通っているのであれば、テストが通ることを判断の根拠とすることはできないと思います。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,19 @@
-  def show; end
+  def show
+    @blog = if user_signed_in?
+              Blog.where(id: params[:id], user: current_user).or(Blog.where(secret: false, id: params[:id])).find_by!(id: params[:id])
```
`id: params[:id]` の指定があちこちに重複しています。また、ID で検索する場合は `find_by!` ではない別の API が使えるはずです。

Ryzhon replied:
条件を整理して簡潔に実装することができました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,19 @@
 
-  def show; end
+  def show
+    @blog = if user_signed_in?
+              Blog.where(id: params[:id], user: current_user).or(Blog.where(secret: false, id: params[:id])).find_by!(id: params[:id])
+            else
+              Blog.where(secret: false).find_by!(id: params[:id])
+            end
```
この条件分岐自体も無くせると思います。無くせたとして、なぜそれで動くのかをコメントしてください。

cafedomancer replied:
上記のコメントを残してください。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -28,7 +35,12 @@ def create
 
   def update
-    if @blog.update(blog_params)
+    updated_params = if !current_user.premium && blog_params[:random_eyecatch] == 'true' && !@blog.random_eyecatch
+                       blog_params.merge(random_eyecatch: false)
```
結局、有料ユーザーでない場合は、元の `@blog.random_eyecatch` の値を維持するようにしているだけで、`false` に設定したいわけではないような気がします。が、`false` に設定しようとしているので、コードからは意図が適切に伝わらくなってしまっていると思います。また、create のアクションと処理が部分的に重複している点も気になります。以上を踏まえて、別の方法を考えてください。

Ryzhon replied:
重複している部分を共通化し、処理を明確にすることでコードから意図を伝わりやすくしました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,19 @@
+    @blog = if user_signed_in?
+              Blog.where(id: params[:id], user: current_user).or(Blog.where(secret: false, id: params[:id])).find_by!(id: params[:id])
+            else
+              Blog.where(secret: false).find_by!(id: params[:id])
```
`where(secret: false)` の部分は既存のスコープが使えるはずです。

Ryzhon committed:
blogs_controllerのshowのロジックを簡略化

Ryzhon committed:
処理を共通化してコードの意図を伝わりやすくする

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -43,8 +47,14 @@ def destroy
+  def adjust_random_eyecatch(blog_params)
+    return blog_params if current_user.premium
+
+    blog_params.merge(random_eyecatch: false)
```
https://github.com/Ryzhon/haunted-blog/pull/1#discussion_r1530101872 にコメントしている問題が解決していないと思うのですが、気のせいですかね？ 考えた対処方法についてアドバイスすることはできますが、プラクティスの課題なので対処方法自体は自分で考えてください。

Ryzhon replied:
ユーザーがプレミアムユーザーでないときはストロングパラメーターにrandom_eyecatchを含めないようにしました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -43,8 +47,14 @@ def destroy
 
-  def set_blog
-    @blog = Blog.find(params[:id])
+  def adjust_random_eyecatch(blog_params)
+    return blog_params if current_user.premium
+
+    blog_params.merge(random_eyecatch: false)
+  end
```
`adjust_random_eyecatch` メソッドは `blog_params` メソッドと同じ箇所でしか使われていないのだから、`blog_params` メソッドそのものを修正したほうがよいと思います。

Ryzhon replied:
ユーザーがプレミアムユーザーでないときはストロングパラメーターにrandom_eyecatchを含めないようにしました。

Ryzhon committed:
ユーザーがプレミアムユーザーでないときはストロングパラメーターにrandom_eyecatchを含めないようにする
詳細：下記のケースに対応
1. 有料ユーザーで登録
2. ランダムアイキャッチを指定したブログを作成
3. 無料ユーザーに変更
4. 1 で作成したブログを更新 -> ランダムアイキャッチが意図せず無効になる

無料ユーザーに変更してからブログを更新してもrandom_eyecatchのパラメータを許していないので、無料ユーザーがブログを変更してもrandom_eyecatchがfalseになることはない

Ryzhon committed:
個人情報であるid, emailのみを削除して、他の情報は表示するようにする

Ryzhon commented on app/controllers/blogs_controller.rb:
```diff
@@ -2,14 +2,15 @@
-  def show; end
+  def show
+    @blog = Blog.where(user: current_user).or(Blog.published).find(params[:id])
+  end
```
ここです。

cafedomancer replied:
条件分岐をなくしてもなぜ動くのかをコメントしてください。

Ryzhon replied:
修正前はuser_signed_inでユーザーがサインインしているかで条件分岐されていました。
しかし、修正後はログインしているユーザー or 公開されているブログからidをもとに探しているため、条件分岐が必要なくなりました。ブログをcurrent_userで探せなかった場合は公開されているブログから探すことができるので、動きます。

cafedomancer replied:
うーん...？ `user_signed_in?` による条件分岐が必要なくなった理由になっていないと思います。別の質問をしますが、`user_signed_in?` による条件分岐を行っていたのはなぜですか？

Ryzhon replied:
`user_signed_in?`による条件分岐を行っていたのは`current_user`が存在するか否か確認するためです。currnet_userはuser_signed_inがtrueだった場合のクエリに必要です。しかし、わかりやすくするためにはuser_signed_inにした方がわかりやすかったと思います。

cafedomancer replied:
元のクエリでも今のクエリでも取得できるデータ集合に変わりはないはずです (以下参照) が、元のクエリで `current_user` が必要だったのだとすると、今のクエリでも `current_user` が必要だということになります。元のクエリで `current_user` が必要な理由か、もしくは今のクエリで `current_user` が必要ない理由を説明してください。

```rb
Blog.where(id: params[:id], user: current_user).or(Blog.where(secret: false, id: params[:id])).find_by!(id: params[:id])
```
```rb
Blog.where(user: current_user).or(Blog.published).find(params[:id])
```

Ryzhon replied:
元のクエリでcurrent_userが必要な理由は、シークレットがtrueの場合もクエリに引っかかるようにしたかったからです。

cafedomancer replied:
質問と回答が噛み合っていないようなので確認内容を変えます。まず、`current_user` が取り得る値をコメントしてください。次に、それらの値に対して `Blog.where(user: current_user).or(Blog.published).find(params[:id])` の発行する SQL クエリが意図したとおりに動く理由を説明してください。

Ryzhon replied:
current_userが取り得る値はユーザーがログインしている場合はUserオブジェクトでログインしていない場合はnilが返されます。`Blog.where(user: current_user).or(Blog.published).find(params[:id])`の発行するSQLクエリが意図した通りに動く理由はif文で書かれていた条件がor文で表現できているからです。

cafedomancer replied:
> current_userが取り得る値はユーザーがログインしている場合はUserオブジェクトでログインしていない場合はnilが返されます。

この部分は合っていますが、

> `Blog.where(user: current_user).or(Blog.published).find(params[:id])`の発行するSQLクエリが意図した通りに動く理由はif文で書かれていた条件がor文で表現できているからです。

この部分は合っていない気がします。SQL の OR キーワードは条件分岐を表すものではなく集合の論理和を表すものだからです。`current_user` の値に応じて発行されるクエリによってどのようなデータ集合が得られるのかを考え、その上で意図どおりに動く理由を改めてコメントしてください。

Ryzhon replied:
意図としてはsecretがtrueのブログはそのブログを所持しているユーザーは見ることができ、その他のユーザーは見ることができないことです。そして、orは論理和を表しているためcurrent_userがUserオブジェクトだった場合はログインしているユーザーが所持しているブログを返し、current_userがnilだった場合はsecretがfalseのブログを取得することができるため、要件を満たしています。

cafedomancer replied:
> そして、orは論理和を表しているためcurrent_userがUserオブジェクトだった場合はログインしているユーザーが所持しているブログを返し、current_userがnilだった場合はsecretがfalseのブログを取得することができるため、要件を満たしています。

それは条件分岐ではないですかね？ 仮にそのような動きをしているのだとして、`current_user` が `User` オブジェクトである (つまりログイン中である) 場合に、`secret` が `false` に設定されている他のユーザーのブログにアクセスしても、ログイン中のそのユーザーはそのようなブログを見られないことなりそうな気がするのですが。要件としてはそのようなブログも見られてよいはずです。挙動に対する認識になにか誤りがありそうなので改めて整理してみてください。

Ryzhon replied:
付け加えます。
orは論理和を表しているためcurrent_userがUserオブジェクトだった場合はログインしているユーザーが所持しているブログに**加えて、公開されているブログ**を返し、current_userがnilだった場合はsecretがfalseのブログを取得することができるため、要件を満たしています。

cafedomancer replied:
> current_userがUserオブジェクトだった場合はログインしているユーザーが所持しているブログに**加えて、公開されているブログ**を返し、

この部分は論理和が表現されていますが、

> current_userがnilだった場合はsecretがfalseのブログを取得することができる

この部分は論理和が表現されていないように見えます。条件分岐ではなく論理和なのだから、`current_user`が `nil` の場合でもユーザーが所持するブログの検索は行われるはずです。この部分についての説明をコメントしてください。

Ryzhon replied:
付け加えます。
orは論理和を表しているためcurrent_userがUserオブジェクトだった場合はログインしているユーザーが所持しているブログに加えて、公開されているブログを返し、current_userがnilだった場合は**ユーザーがないため、ユーザーが所持するブログを取得できずに**secretがfalseのブログのみが検索されるため、要件を満たしています。

cafedomancer replied:
「ユーザーが所持するブログを取得できずに」の「取得できずに」という部分が曖昧なので確認したいのですが、具体的にはどういうことを指していますか？ (エラーが起こるなど)

Ryzhon replied:
エラーは起こりません。
current_userがnilなので`Blog.where(user: current_user)`の部分は集合に含まれず、それ以降の`or(Blog.published)`から取得されます。

cafedomancer replied:
OK ですー。改めて整理すると、ログイン済みであるかどうかに応じて `current_user` の値が `User` オブジェクトか `nil` のどちらかになり、それに応じてそれぞれ以下のようなクエリの発行とデータの取得が行われるので、上記のコードで期待する動作になります。

ログイン済みである場合 (`current_user` が `User` オブジェクト):

```sql
SELECT * FROM blogs WHERE (user_id = <current_user.id> OR published = TRUE) AND id = <params[:id]>;
```

- `user_id = <current_user.id>` の部分で取得されるデータ: ログイン中のユーザーのブログ
- `published = TRUE` の部分で取得されるデータ: 公開設定になっているブログ (自分のものも他人のものも含む)
- -> 結果として、ログイン中のユーザーとして見てよいブログがデータとして得られる

ログイン済みでない (`current_user` が `nil`):

```sql
SELECT * FROM blogs WHERE (user_id IS NULL OR published = TRUE) AND id = <params[:id]>;
```

- `user_id IS NULL` の部分で取得されるデータ: `blogs.user_id` カラムには NOT NULL 制約がかかっており、`user_id` が `NULL` であるようなデータは存在しないため無し
- `published = TRUE` の部分で取得されるデータ: 公開設定になっているブログ (自分のものも他人のものも含む) 
- -> 結果として、公開設定になっているブログ**のみ**がデータとして得られる

Ryzhon replied:
次回からは期待される挙動に加えて、条件の詳細とそれで行われる実際の挙動を説明するようにします。
ありがとうございました。

Ryzhon commented:
プラクティスの要件を読み直しました。
個人情報であるid, emailだけを削除しました。
[最後のコメント](https://github.com/Ryzhon/haunted-blog/pull/1#discussion_r1536713463)については[こちら](https://github.com/Ryzhon/haunted-blog/pull/1#discussion_r1538716003)修正済みです。ご確認よろしくお願いいたします。

cafedomancer commented on app/views/api/liking_users/_user.json.jbuilder:
```diff
@@ -1 +1 @@
-json.extract! user, :id, :email, :created_at, :updated_at, :nickname, :premium
+json.extract! user, :created_at, :updated_at, :nickname, :premium
```
これだと最初にコメントしている「不要な情報のみを削るよりも、必要な情報のみを吐くほうが考え方としてはより安全です。」という内容が実現できていないです。それに必要な情報も引き続き削除されてしまっています。必要な情報のみを出力するようにし、それがなぜ必要なのかを根拠とともに示してください。

Ryzhon replied:
id, nicknameのみを表示するようにしました。
_user.json.jbuilderはliking_users_controller.rbのindexでソートするためにidが使われています。
nicknameはLinkingApp.jsxでライクした人を表示するために使われているためidとnicknameのみを残しました。

cafedomancer replied:
> nicknameはLinkingApp.jsxでライクした人を表示するために使われているためidとnicknameのみを残しました。

この部分は正しいですが、

> _user.json.jbuilderはliking_users_controller.rbのindexでソートするためにidが使われています。

この部分は正しくないです。これは何の説明をしているんでしょうか？

Ryzhon replied:
勘違いしていました。
_user.json.jbuilderでidを削除するとliking_users_controller.rbの`@users = blog.liking_users.order(:id)`でidがないため並べ替えることができないと思っていました。

正しくはLinkingApp.jsx内でidは使われているために必要です。

cafedomancer replied:
わかりましたー。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -2,14 +2,15 @@
 
 class BlogsController < ApplicationController
   skip_before_action :authenticate_user!, only: %i[index show]
-
```
終了条件を満たすのに必要ではない変更は含めないでください。他にも必要のない変更が含まれているので修正してください。

Ryzhon replied:
mainと比較して不要な変更をなくしました。

cafedomancer replied:
修正されていないです。

Ryzhon replied:
どこの部分を言っていますか？
showの変更はバグの改善に必要なはずです。

cafedomancer replied:
5 行目の空行の削除についてコメントしています。

Ryzhon replied:
空行を削除しました。

Ryzhon committed:
mainと比較して不要な変更をなくす

Ryzhon committed:
id, nicknameのみを表示するようにしました。
_user.json.jbuilderはliking_users_controller.rbのindexでソートするためにidが使われています。
nicknameはLinkingApp.jsxでライクした人を表示するために使われているためidとnicknameのみを残しました。

Ryzhon committed:
mainブランチと比較したときの不要な変更を削除
詳細：空行を削除
